android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiToolbar.java

```diff
@@ -56,14 +63,32 @@
 	private final int TOOLBAR_SET_CONTENT_INSETS_RELATIVE = 10018;
 	//endregion
 
+	private Map<String, TiDrawableReference> drawableMap = new HashMap<String, TiDrawableReference>();
+	private KrollFunction resourceLoadedListener = null;
 	/**
 	 * Constructs a TiUIView object with the associated proxy.
 	 * @param proxy the associated proxy.
 	 * @module.api
 	 */
-	public TiToolbar(TiViewProxy proxy) {
+	public TiToolbar(final TiViewProxy proxy) {
 		super(proxy);
-		toolbar = new Toolbar(proxy.getActivity());
+		toolbar = new Toolbar(proxy.getActivity()) {
+
+			// Check the resource loading once the view is drawn
+			@Override
+			protected void onDraw(Canvas canvas) {
+				super.onDraw(canvas);
+				//if we have attached listener send the dictionary
+				if (resourceLoadedListener != null) {
+					KrollDict callbackParam = new KrollDict();
+					callbackParam.put(TiC.PROPERTY_LOGO, checkResourceLoading(drawableMap.get(TiC.PROPERTY_LOGO), this.getLogo()));
+					callbackParam.put(TiC.PROPERTY_OVERFLOW_ICON, checkResourceLoading(drawableMap.get(TiC.PROPERTY_OVERFLOW_ICON), this.getOverflowIcon()));
+					callbackParam.put(TiC.PROPERTY_NAVIGATION_ICON, checkResourceLoading(drawableMap.get(TiC.PROPERTY_NAVIGATION_ICON), this.getNavigationIcon()));
```

Comment:
The checkResourceLoading() method compares images via Bitmap.sameAs(), which is a per-pixel comparison. I'm worried about performance since this code is doing this on every onDraw(). So, I'm guessing you just want to know when the image/drawable is done loading, right?

Reply:
Not only when it was loaded, but also if it was shown correctly. That was the problem that required these checks - it could not get a reference to the source when it is placed in the density specific folders and was passing `TiDrawableReference` that was 'empty' (if I can put it that way). It was loaded successfully but had not actual content.
I was not able to come up with anything else that could serve this purpose.

Reply:
> It was loaded successfully but had not actual content

Just to be clear, images in the toolbar are being loaded fine, right? The issue is that the callback you're implementing was being invoked before the image was loaded or rendered onscreen?

I'm a bit confused by this. `TiDrawableReference.getDrawable()` will return a BitmapDrawable **after** the image was fully decoded and loaded as a bitmap and it's a blocking call (not async). I don't see how this can go wrong. But I can see the BitmapDrawable being rendered after the callback was invoked. Perhaps you're missing an invalidate() or requestLayout() call after adding views or assigning drawables?

Reply:
They were not loaded if they were placed in a density specific folder. `TiDrawableReference.getDrawable()` returned `null` in that case which was passed to `Toolbar.setLogo()`. And the idea of this callback is to be used only for testing whether an existing resource was properly loaded as a logo/overflowIcon/navigationIcon. I added it in the `onDraw()` method, because I think that way it will catch any problems with the image, not only if the `getDrawable()` is `null`. I don't imagine it has any use outside of testing purposes. Maybe we chat about that if it does not make sense?

Reply:
The `TiDrawableReference.getDrawable()` can definitely load density specific images from the res directory (ex: res/drawable-xhdpi, res/drawable-xxhdpi, etc.). I've done it. The problem is that this is difficult to test via our "TitaniumTest" Android project because Titanium expects the "res" images to be posted-fixed with an MD5 hash. It's not obvious and I think it makes it overly complicated, but the idea was to avoid filename collision under the APK's res folder. I find it easier to test it via Appcelerator Studio since our "appc" build system will automatically copy images under the "res" folder and postfix them with the expected MD5 hash.

Try this. Create a "Classic" app in Appcelerator Studio and then put an image file under the following directories...
`./Resources/android/images/res-mdpi/<YourImage.png>`
`./Resources/android/images/res-hdpi/<YourImage.png>`

Our appc build will copy the above image under the appropriate res folder when generating a Android project under the "build/android" folder. You can access the above density scaled image using the following path in Titanium...
`/images/<YourImage.png>`

Reply:
It is working now, yes. We had a regression in 6.2.0 for that which was fixed. The main point of the PR is to add tests to prevent it from happening again. The regression affected 3 places - Ti.UI.ImageView, Toolbar logo/overflowIcon/navigationIcon and BigStyleNotification image. I don't think we can have mocha tests for the latter, because they are visible out of JavaScript context. The test for ImageView relies on the `load` event, but for Toolbar's properties that require drawables that was the only reliable lifecycle moment I was able to come up with.
I went through the MD5 hash adding and have successfully reproduced it in TitaniumTest and debug it there. Also I have tried the tests with only this change applied on 6.2.0. The purpose of this event there is entirely for testing and is not expected to be used for anything else.